src/Orckestra.Composer.Website/App_Insights/AppInsightsListener.cs

```diff
@@ -0,0 +1,155 @@
+                    "ProcessId", 
+                    "Process Name", 
+                    "Thread Name", 
+                    "Win32 ThreadId"
+                };
+
+                foreach (var el in propertiesToProcess)
+                {
+                    AddTelemetryExceptionProperty(message, el, telemetryExceptionProperties);
+                }
+
+                var exception = new ExceptionTelemetry(
+                    new List<ExceptionDetailsInfo>() { exceptionDetailsInfo }, 
+                    severityLevel, 
+                    null,
+                    telemetryExceptionProperties, 
+                    new Dictionary<string, double>());
+
+                telemetryClient.TrackException(exception);
+            }
+        }
+
+        private bool IsC1FunctionException(string message, out string functionName)
+        {
+            functionName = null;
+            Regex regex = new Regex("^Title:.+? Function: (.+?)$", RegexOptions.Multiline | RegexOptions.Compiled | RegexOptions.IgnoreCase);
+            var match = regex.Match(message);
+            if (!match.Success) return false;
+
+            functionName = match.Groups[1].Value.Trim();
+            return true;
+        }
+
+        private SeverityLevel GetSeverityLevel(string message)
+        {
+            SeverityLevel severityLevel = SeverityLevel.Error;
+            Regex regex = new Regex($"Severity: (.+?)$", RegexOptions.Multiline | RegexOptions.Compiled | RegexOptions.IgnoreCase);
+            Match match = regex.Match(message);
+            if (!match.Success) return severityLevel;
+
+            switch (match.Groups[1].Value.Trim())
+            {
+                case "Fine":
+                case "Info":
+                    severityLevel = SeverityLevel.Information;
+                    break;
+                case "Fatal":
+                    severityLevel = SeverityLevel.Critical;
+                    break;
+                case "Warning":
+                    severityLevel = SeverityLevel.Warning;
+                    break;
+                case "Debug":
+                    severityLevel = SeverityLevel.Verbose;
+                    break;
+                case "Error":
+                    severityLevel = SeverityLevel.Error;
+                    break;
+            }
+            return severityLevel;
+        }
+
+        private string GetExceptionTypeName(string message)
+        {
+            Regex regex = new Regex("^Message: (.+?):", RegexOptions.Multiline | RegexOptions.Compiled | RegexOptions.IgnoreCase);
+            Match match = regex.Match(message);
+            return !match.Success ? null : match.Groups[1].Value.Trim();
+        }
+
+        private string GetCallStackContent(string message)
+        {
+            var keyWord = "Extended Properties:";
+            var stackIndex = message.IndexOf(keyWord);
+            return stackIndex < 0 ? null : message.Remove(0, stackIndex + keyWord.Length).Trim();
```

Comment:
With async code the  call stacks could be rather huge. Both to reduce potential storage costs on AppInsights side, and to improve readability, it would make sense to filter out lines like these:

```
const string[] CallstackLinesToIgnore = {
"--- End of stack trace from previous location where exception was thrown ---",
"at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()",
"at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)",
"at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()",
"at System.Threading.Tasks.Task.Execute()"
}
```

An example of a call stack:

```
Exception - System.AggregateException: One or more errors occurred. ---> System.UriFormatException: Invalid URI: The hostname could not be parsed.
   at Orckestra.Overture.RestClient.OvertureClient.<>c__DisplayClass46_0`1.<SendAsync>b__0(Task`1 t)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.ComposerOvertureClient.<InterceptAsync>d__18`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.Search.Repositories.SearchRepository.<ExecuteProductSearchRequestAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.Search.Repositories.SearchRepository.<SearchProductAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.Search.Services.BaseSearchViewService`1.<SearchAsync>d__45.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.Search.Services.SearchViewService.<GetProductSearchResultsAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.Search.Services.SearchViewService.<GetSearchViewModelAsync>d__4.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at ASP._Page_App_Data_Razor_Composer_Search_ContentTabs_cshtml.get_ProductsSearchViewModel() in D:\BetterRetail\deployment\Website\App_Data\Razor\Composer\Search\ContentTabs.cshtml:line 33
   at ASP._Page_App_Data_Razor_Composer_Search_ContentTabs_cshtml.get_ProductSearchResults() in D:\BetterRetail\deployment\Website\App_Data\Razor\Composer\Search\ContentTabs.cshtml:line 34
   at ASP._Page_App_Data_Razor_Composer_Search_ContentTabs_cshtml.IsProductsCorrected() in D:\BetterRetail\deployment\Website\App_Data\Razor\Composer\Search\ContentTabs.cshtml:line 38
   at ASP._Page_App_Data_Razor_Composer_Search_ContentTabs_cshtml.Execute() in D:\BetterRetail\deployment\Website\App_Data\Razor\Composer\Search\ContentTabs.cshtml:line 59
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy()
   at System.Web.WebPages.WebPage.ExecutePageHierarchy(IEnumerable`1 executors)
   at System.Web.WebPages.WebPage.ExecutePageHierarchy()
   at Composite.AspNet.Razor.CompositeC1WebPage.ExecutePageHierarchy()
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy(WebPageContext pageContext, TextWriter writer, WebPageRenderingBase startPage)
   at Composite.AspNet.Razor.RazorHelper.ExecuteRazorPage(WebPageBase webPage, Action`1 setParameters, Type resultType, FunctionContextContainer functionContextContainer)
   at Composite.AspNet.Razor.RazorHelper.ExecuteRazorPage(String virtualPath, Action`1 setParameters, Type resultType, FunctionContextContainer functionContextContainer)
   at Composite.Plugins.Functions.FunctionProviders.RazorFunctionProvider.RazorBasedFunction.Execute(ParameterList parameters, FunctionContextContainer context)
   at Composite.Functions.Foundation.PluginFacades.FunctionWrapper.Execute(ParameterList parameters, FunctionContextContainer context)
---> (Inner Exception #0) System.UriFormatException: Invalid URI: The hostname could not be parsed.
   at Orckestra.Overture.RestClient.OvertureClient.<>c__DisplayClass46_0`1.<SendAsync>b__0(Task`1 t)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.ComposerOvertureClient.<InterceptAsync>d__18`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.Search.Repositories.SearchRepository.<ExecuteProductSearchRequestAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.Search.Repositories.SearchRepository.<SearchProductAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.Search.Services.BaseSearchViewService`1.<SearchAsync>d__45.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.Search.Services.SearchViewService.<GetProductSearchResultsAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orckestra.Composer.Search.Services.SearchViewService.<GetSearchViewModelAsync>d__4.MoveNext()<---
```